kernel/exec.c

```diff
@@ -122,99 +158,144 @@ execelf(const char *progname, const char *path, const char* const *argv,
   oldpgdir = curproc->pgdir;
   curproc->pgdir = pgdir;
   curproc->sz = sz;
-  curproc->tf->eip = elf->entry;  // main
+  curproc->tf->eip = elf.entry;  // main
   curproc->tf->esp = sp;
   switchuvm(curproc);
   freevm(oldpgdir);
-  *ppgdir = 0;  // Clear pgdir to avoid to be free
+
   return 0;
 
-bad:
-  *ppgdir = pgdir;
+ bad:
+  if( pgdir != 0 )
+    freevm(pgdir);
+  if( ip != 0 ){
+    iunlockput(ip);
+    end_op();
+  }
   return -1;
 }
 
-int
-execshebang(const char *path, const char * const *argv,
```

Comment:
Introducing environment variables changes logic in exec ( strictly speaking, exec in xv6 should be called execv in POSIX). So I keep function name only.

Reply:
FYI, in addition to this, there is a race condition between checking a binary type and execution.
To avoid this race, we should lock inode for path from execve  to execelf or check binary type for each binary handler. 
However, as far as I know,  POSIX does not require this because this might cause tremendous performance issue when we meet cascaded shebang cases.

Reply:
On second thought, I think that execve should be written as follows at least to keep the compatibility to your codes:
```
-    if ( rc != 0 )
-      goto error_out;
-  } else if ( ( rd_size >= EXECVE_SHEBANG_STRLEN ) &&
+    if ( rc == 0 )
+      goto out;
+  }
+
+  if ( ( rd_size >= EXECVE_SHEBANG_STRLEN ) &&
     ( memcmp((void *)&exeblk[0], EXECVE_SHEBANG_STR , 2) == 0 ) ) {

     rc = execshebang(path, argv,envp);
-    if ( rc != 0 )
-      goto error_out;
+    if ( rc == 0 )
+      goto out;
   }
-  else
-    panic("execve: unknwon binary");

+  panic("execve: unknwon binary");
+
```
But I'm not sure ilock handling in your codes were written intentionally  or not.
I’m sorry to bother you, would you please notice that again, if you want lock inode between exec and execelf and execshebang.

Reply:
Hi,

After all, I found the intention of your code.
In my understanding, You would like to make much account to consumption of kernel stack.

Reply:
I've checked POSIX ( SUSv4 ), there is no description about mutual exclusion of file access during exec family. So I think both your code and mine might be correct. 

BTW, new POSIX specification (2017) defines arguments in main function more precisely like:
```
Some implementations provide a third argument to main() called envp. This is defined as a pointer to the environment. The ISO C standard specifies invoking main() with two arguments, so implementations must support applications written this way. Since this volume of POSIX.1-2017 defines the global variable environ, which is also provided by historical implementations and can be used anywhere that envp could be used, there is no functional need for the envp argument. Applications should use the getenv() function rather than accessing the environment directly via either envp or environ. Implementations are required to support the two-argument calling sequence, but this does not prohibit an implementation from supporting envp as an optional third argument.
```
According to this definition, we should treat the environment array as the third argument of main function from _start, but the C prototype declaration for main function should have two argument only ( As prestigious man in UNIX, Mr. Pate said ).